Operator secrets sequencing

### Before reporting an issue

- [X] I have searched existing issues
- [X] I have reproduced the issue with the [latest nightly release](https://github.com/keycloak/keycloak/releases/tag/nightly)

### Area

operator

### Describe the bug

When deploying using the operator the operator checks if the db secrets exist and if they do not then it stops further deployment.

If the secrets are later provisioned this does not get picked up by the operator so the deployment becomes stuck.

Successful deployment requires the secrets are in place prior to creating the keycloak definition. This makes an implied sequencing.

Ideally the operator will retry the creation of the deployment/statefulset or it will create the stateful regardless of the secret and allow the sts and pod operations to implement the retries.


### Version

22.0.0

### Expected behavior

Creation of the stateless set occurs regardless of the presence of the db credentials secret or the operator checks for presence of secrets after a timeout

### Actual behavior

After the initial check for secrets when the keycloak object is created and faulted no further updates or progress is made.

### How to Reproduce?

Create a keycloak definition using secret references to db username and password referring to a secret that does not exist.
The stateful set will not be created.
Create the appropriate secret.
The operator will not update.

Now delete the keycloak object then recreate.
The stateful set is now constructed. Showing that the operator has an implied sequencing that should ideally be removed or avoided to ensure consistent deployments

### Anything else?

_No response_

I have tried this a few times and I am finding it very easy to create this error using the following sequence:

Follow the notes here: https://www.keycloak.org/operator/basic-deployment
But do not create the secret for the db. (keycloak-db-secret)

then create the Keycloak definition as below.

`apiVersion: k8s.keycloak.org/v2alpha1
kind: Keycloak
metadata:
  name: example-kc
spec:
  instances: 1
  db:
    vendor: postgres
    host: postgres-db
    usernameSecret:
      name: keycloak-db-secret
      key: username
    passwordSecret:
      name: keycloak-db-secret
      key: password
  http:
    httpPort: 8080
    httpEnabled: true
  hostname:
    strict: false
  ingress:
      enabled: false
`

If you watch the operator as this is applied you can see it failing with a java stack error as a 404 trying to access the secret:

`Caused by: io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: GET at: https://10.96.0.1:443/api/v1/namespaces/keycloak/secrets/keycloak-db-secret. Message: secrets "keycloak-db-secret" not found. Received status: Status(apiVersion=v1, code=404, details=StatusDetails(causes=[], group=null, kind=secrets, name=keycloak-db-secret, retryAfterSeconds=null, uid=null, additionalProperties={}), kind=Status, message=secrets "keycloak-db-secret" not found, metadata=ListMeta(_continue=null, remainingItemCount=null, resourceVersion=null, selfLink=null, additionalProperties={}), reason=NotFound, status=Failure, additionalProperties={}).
`

As you described @shawkins this does repeat and retry after (not 10 seconds but much lower). BUT it does not continue to do this and fails permanently after a set number of tries (not sure exactly how many.... looks to be approx 5-10).

After that point it no longer checks for the secrets.
From creation to no longer trying takes: about 20secs
ie from `2023-08-12 19:45:45,938` until `2023-08-12 19:46:04,601`

If I now create the relevant secret on the cluster.
`kubectl -n keycloak create secret generic keycloak-db-secret --from-literal=username=keycloak --from-literal=password=pgpw1`
Nothing happens as the operator has stopped watching.

This makes the operator dependant upon the order of the secret creation and therefore not idempotent.
As a result this requires that the operator must use a sequencing around it to make it work correctly.

Hi Team, I was able to reproduce the issue.

The installation flow is
1. Create KeyCloak CR
```
apiVersion: k8s.keycloak.org/v2alpha1
kind: Keycloak
metadata:
  name: cs-keycloak
  namespace: keycloak-integration
spec:
  db:
    host: keycloak-edb-cluster-rw
    passwordSecret:
      key: password
      name: keycloak-edb-cluster-app
    usernameSecret:
      key: username
      name: keycloak-edb-cluster-app
    vendor: postgres
  hostname:
    strict: false
  http:
    tlsSecret: cs-ca-certificate-secret
  ingress:
    className: openshift-default
    enabled: true
  instances: 1
```
2. Create EDB postgreSQL `Cluster` Kind CR to provision KeyCloak.
```

apiVersion: postgresql.k8s.enterprisedb.io/v1
kind: Cluster
metadata:
  name: keycloak-edb-cluster
  namespace: keycloak-integration
spec:
    bootstrap:
      initdb:
        database: keycloak
        owner: app
    instances: 1
    logLevel: info
    primaryUpdateStrategy: unsupervised
    storage:
      size: 1Gi
    walStorage:
      size: 1Gi
```

When KeyCloak Operator starts to reconcile the CR, it raise error on missing secret `keycloak-edb-cluster-app`.
```
2023-08-23 18:07:05,181 ERROR [io.jav.ope.pro.eve.ReconciliationDispatcher] (ReconcilerExecutor-keycloakcontroller-72) Error during event processing ExecutionScope{ resource id: ResourceID{name='cs-keycloak', namespace='keycloak-integration'}, version: 159092995} failed.: io.fabric8.kubernetes.client.ResourceNotFoundException: Resource couldn't be fetched : Failure executing: GET at: https://172.30.0.1:443/api/v1/namespaces/keycloak-integration/secrets/keycloak-edb-cluster-app. Message: secrets "keycloak-edb-cluster-app" not found. Received status: Status(apiVersion=v1, code=404, details=StatusDetails(causes=[], group=null, kind=secrets, name=keycloak-edb-cluster-app, retryAfterSeconds=null, uid=null, additionalProperties={}), kind=Status, message=secrets "keycloak-edb-cluster-app" not found, metadata=ListMeta(_continue=null, remainingItemCount=null, resourceVersion=null, selfLink=null, additionalProperties={}), reason=NotFound, status=Failure, additionalProperties={}).
	at io.fabric8.kubernetes.client.dsl.internal.BaseOperation.require(BaseOperation.java:155)
	at io.fabric8.kubernetes.client.dsl.internal.BaseOperation.require(BaseOperation.java:92)
	at org.keycloak.operator.controllers.WatchedSecretsStore.lambda$fetchCurrentSecrets$3(WatchedSecretsStore.java:158)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.HashMap$KeySpliterator.forEachRemaining(HashMap.java:1707)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)
	at org.keycloak.operator.controllers.WatchedSecretsStore.fetchCurrentSecrets(WatchedSecretsStore.java:159)
	at org.keycloak.operator.controllers.WatchedSecretsStore.<init>(WatchedSecretsStore.java:70)
	at org.keycloak.operator.controllers.KeycloakController.reconcile(KeycloakController.java:110)
	at org.keycloak.operator.controllers.KeycloakController.reconcile(KeycloakController.java:49)
	at org.keycloak.operator.controllers.KeycloakController_ClientProxy.reconcile(Unknown Source)
	at io.javaoperatorsdk.operator.processing.Controller$1.execute(Controller.java:152)
	at io.javaoperatorsdk.operator.processing.Controller$1.execute(Controller.java:110)
	at io.javaoperatorsdk.operator.api.monitoring.Metrics.timeControllerExecution(Metrics.java:219)
	at io.javaoperatorsdk.operator.processing.Controller.reconcile(Controller.java:109)
	at io.javaoperatorsdk.operator.processing.event.ReconciliationDispatcher.reconcileExecution(ReconciliationDispatcher.java:140)
	at io.javaoperatorsdk.operator.processing.event.ReconciliationDispatcher.handleReconcile(ReconciliationDispatcher.java:121)
	at io.javaoperatorsdk.operator.processing.event.ReconciliationDispatcher.handleDispatch(ReconciliationDispatcher.java:91)
	at io.javaoperatorsdk.operator.processing.event.ReconciliationDispatcher.handleExecution(ReconciliationDispatcher.java:64)
	at io.javaoperatorsdk.operator.processing.event.EventProcessor$ReconcilerExecutor.run(EventProcessor.java:409)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: GET at: https://172.30.0.1:443/api/v1/namespaces/keycloak-integration/secrets/keycloak-edb-cluster-app. Message: secrets "keycloak-edb-cluster-app" not found. Received status: Status(apiVersion=v1, code=404, details=StatusDetails(causes=[], group=null, kind=secrets, name=keycloak-edb-cluster-app, retryAfterSeconds=null, uid=null, additionalProperties={}), kind=Status, message=secrets "keycloak-edb-cluster-app" not found, metadata=ListMeta(_continue=null, remainingItemCount=null, resourceVersion=null, selfLink=null, additionalProperties={}), reason=NotFound, status=Failure, additionalProperties={}).
	at io.fabric8.kubernetes.client.KubernetesClientException.copyAsCause(KubernetesClientException.java:238)
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.waitForResult(OperationSupport.java:518)
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.handleResponse(OperationSupport.java:535)
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.handleGet(OperationSupport.java:478)
	at io.fabric8.kubernetes.client.dsl.internal.BaseOperation.handleGet(BaseOperation.java:741)
	at io.fabric8.kubernetes.client.dsl.internal.BaseOperation.requireFromServer(BaseOperation.java:185)
	at io.fabric8.kubernetes.client.dsl.internal.BaseOperation.require(BaseOperation.java:153)
	... 26 more
Caused by: io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: GET at: https://172.30.0.1:443/api/v1/namespaces/keycloak-integration/secrets/keycloak-edb-cluster-app. Message: secrets "keycloak-edb-cluster-app" not found. Received status: Status(apiVersion=v1, code=404, details=StatusDetails(causes=[], group=null, kind=secrets, name=keycloak-edb-cluster-app, retryAfterSeconds=null, uid=null, additionalProperties={}), kind=Status, message=secrets "keycloak-edb-cluster-app" not found, metadata=ListMeta(_continue=null, remainingItemCount=null, resourceVersion=null, selfLink=null, additionalProperties={}), reason=NotFound, status=Failure, additionalProperties={}).
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.requestFailure(OperationSupport.java:671)
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.requestFailure(OperationSupport.java:651)
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.assertResponseCode(OperationSupport.java:600)
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.lambda$handleResponse$0(OperationSupport.java:560)
	at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:646)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2147)
	at io.fabric8.kubernetes.client.http.StandardHttpClient.lambda$completeOrCancel$10(StandardHttpClient.java:140)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863)
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:841)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2147)
	at io.fabric8.kubernetes.client.http.ByteArrayBodyHandler.onBodyDone(ByteArrayBodyHandler.java:52)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863)
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:841)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2147)
	at io.fabric8.kubernetes.client.vertx.VertxHttpRequest.lambda$null$1(VertxHttpRequest.java:122)
	at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:264)
	at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:246)
	at io.vertx.core.http.impl.HttpEventHandler.handleEnd(HttpEventHandler.java:76)
	at io.vertx.core.http.impl.HttpClientResponseImpl.handleEnd(HttpClientResponseImpl.java:250)
	at io.vertx.core.http.impl.Http1xClientConnection$StreamImpl.lambda$new$0(Http1xClientConnection.java:444)
	at io.vertx.core.streams.impl.InboundBuffer.handleEvent(InboundBuffer.java:255)
	at io.vertx.core.streams.impl.InboundBuffer.write(InboundBuffer.java:134)
	at io.vertx.core.http.impl.Http1xClientConnection$StreamImpl.handleEnd(Http1xClientConnection.java:708)
	at io.vertx.core.impl.EventLoopContext.lambda$execute$2(EventLoopContext.java:78)
	at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:174)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:167)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:470)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:569)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	... 1 more

2023-08-23 18:07:05,182 ERROR [io.jav.ope.pro.eve.EventProcessor] (ReconcilerExecutor-keycloakcontroller-72) Exhausted retries for ExecutionScope{ resource id: ResourceID{name='cs-keycloak', namespace='keycloak-integration'}, version: 159092995}
```

But it does not re-queue the KeyCloak CR reconciliation when this error happens.
When the secret shows up in the namespace, KeyCloak operator stays silent and does not proceed. So I have to manually restart the operator pod to trigger a reconciliation

I have found that updating a dummy environment var on keycloak object is
best way to trigger operator to re-process.
Not ideal but works for temporary until get a better option in place.


On Thu, 24 Aug 2023, 19:58 Daniel Fan, ***@***.***> wrote:

> Hi Team, I was able to reproduce the issue.
>
> The installation flow is
>
>    1. Create KeyCloak CR
>
> apiVersion: k8s.keycloak.org/v2alpha1
> kind: Keycloak
> metadata:
>   name: cs-keycloak
>   namespace: keycloak-integration
> spec:
>   db:
>     host: keycloak-edb-cluster-rw
>     passwordSecret:
>       key: password
>       name: keycloak-edb-cluster-app
>     usernameSecret:
>       key: username
>       name: keycloak-edb-cluster-app
>     vendor: postgres
>   hostname:
>     strict: false
>   http:
>     tlsSecret: cs-ca-certificate-secret
>   ingress:
>     className: openshift-default
>     enabled: true
>   instances: 1
>
>
>    1. Create EDB postgreSQL Cluster Kind CR to provision KeyCloak.
>
>
> apiVersion: postgresql.k8s.enterprisedb.io/v1
> kind: Cluster
> metadata:
>   name: keycloak-edb-cluster
>   namespace: keycloak-integration
> spec:
>     bootstrap:
>       initdb:
>         database: keycloak
>         owner: app
>     instances: 1
>     logLevel: info
>     primaryUpdateStrategy: unsupervised
>     storage:
>       size: 1Gi
>     walStorage:
>       size: 1Gi
>
> When KeyCloak Operator starts to reconcile the CR, it raise error on
> missing secret keycloak-edb-cluster-app.
>
> 2023-08-23 18:07:05,181 ERROR [io.jav.ope.pro.eve.ReconciliationDispatcher] (ReconcilerExecutor-keycloakcontroller-72) Error during event processing ExecutionScope{ resource id: ResourceID{name='cs-keycloak', namespace='keycloak-integration'}, version: 159092995} failed.: io.fabric8.kubernetes.client.ResourceNotFoundException: Resource couldn't be fetched : Failure executing: GET at: https://172.30.0.1:443/api/v1/namespaces/keycloak-integration/secrets/keycloak-edb-cluster-app. Message: secrets "keycloak-edb-cluster-app" not found. Received status: Status(apiVersion=v1, code=404, details=StatusDetails(causes=[], group=null, kind=secrets, name=keycloak-edb-cluster-app, retryAfterSeconds=null, uid=null, additionalProperties={}), kind=Status, message=secrets "keycloak-edb-cluster-app" not found, metadata=ListMeta(_continue=null, remainingItemCount=null, resourceVersion=null, selfLink=null, additionalProperties={}), reason=NotFound, status=Failure, additionalProperties={}).
> 	at io.fabric8.kubernetes.client.dsl.internal.BaseOperation.require(BaseOperation.java:155)
> 	at io.fabric8.kubernetes.client.dsl.internal.BaseOperation.require(BaseOperation.java:92)
> 	at org.keycloak.operator.controllers.WatchedSecretsStore.lambda$fetchCurrentSecrets$3(WatchedSecretsStore.java:158)
> 	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
> 	at java.base/java.util.HashMap$KeySpliterator.forEachRemaining(HashMap.java:1707)
> 	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
> 	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
> 	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
> 	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
> 	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)
> 	at org.keycloak.operator.controllers.WatchedSecretsStore.fetchCurrentSecrets(WatchedSecretsStore.java:159)
> 	at org.keycloak.operator.controllers.WatchedSecretsStore.<init>(WatchedSecretsStore.java:70)
> 	at org.keycloak.operator.controllers.KeycloakController.reconcile(KeycloakController.java:110)
> 	at org.keycloak.operator.controllers.KeycloakController.reconcile(KeycloakController.java:49)
> 	at org.keycloak.operator.controllers.KeycloakController_ClientProxy.reconcile(Unknown Source)
> 	at io.javaoperatorsdk.operator.processing.Controller$1.execute(Controller.java:152)
> 	at io.javaoperatorsdk.operator.processing.Controller$1.execute(Controller.java:110)
> 	at io.javaoperatorsdk.operator.api.monitoring.Metrics.timeControllerExecution(Metrics.java:219)
> 	at io.javaoperatorsdk.operator.processing.Controller.reconcile(Controller.java:109)
> 	at io.javaoperatorsdk.operator.processing.event.ReconciliationDispatcher.reconcileExecution(ReconciliationDispatcher.java:140)
> 	at io.javaoperatorsdk.operator.processing.event.ReconciliationDispatcher.handleReconcile(ReconciliationDispatcher.java:121)
> 	at io.javaoperatorsdk.operator.processing.event.ReconciliationDispatcher.handleDispatch(ReconciliationDispatcher.java:91)
> 	at io.javaoperatorsdk.operator.processing.event.ReconciliationDispatcher.handleExecution(ReconciliationDispatcher.java:64)
> 	at io.javaoperatorsdk.operator.processing.event.EventProcessor$ReconcilerExecutor.run(EventProcessor.java:409)
> 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
> 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
> 	at java.base/java.lang.Thread.run(Thread.java:833)
> Caused by: io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: GET at: https://172.30.0.1:443/api/v1/namespaces/keycloak-integration/secrets/keycloak-edb-cluster-app. Message: secrets "keycloak-edb-cluster-app" not found. Received status: Status(apiVersion=v1, code=404, details=StatusDetails(causes=[], group=null, kind=secrets, name=keycloak-edb-cluster-app, retryAfterSeconds=null, uid=null, additionalProperties={}), kind=Status, message=secrets "keycloak-edb-cluster-app" not found, metadata=ListMeta(_continue=null, remainingItemCount=null, resourceVersion=null, selfLink=null, additionalProperties={}), reason=NotFound, status=Failure, additionalProperties={}).
> 	at io.fabric8.kubernetes.client.KubernetesClientException.copyAsCause(KubernetesClientException.java:238)
> 	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.waitForResult(OperationSupport.java:518)
> 	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.handleResponse(OperationSupport.java:535)
> 	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.handleGet(OperationSupport.java:478)
> 	at io.fabric8.kubernetes.client.dsl.internal.BaseOperation.handleGet(BaseOperation.java:741)
> 	at io.fabric8.kubernetes.client.dsl.internal.BaseOperation.requireFromServer(BaseOperation.java:185)
> 	at io.fabric8.kubernetes.client.dsl.internal.BaseOperation.require(BaseOperation.java:153)
> 	... 26 more
> Caused by: io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: GET at: https://172.30.0.1:443/api/v1/namespaces/keycloak-integration/secrets/keycloak-edb-cluster-app. Message: secrets "keycloak-edb-cluster-app" not found. Received status: Status(apiVersion=v1, code=404, details=StatusDetails(causes=[], group=null, kind=secrets, name=keycloak-edb-cluster-app, retryAfterSeconds=null, uid=null, additionalProperties={}), kind=Status, message=secrets "keycloak-edb-cluster-app" not found, metadata=ListMeta(_continue=null, remainingItemCount=null, resourceVersion=null, selfLink=null, additionalProperties={}), reason=NotFound, status=Failure, additionalProperties={}).
> 	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.requestFailure(OperationSupport.java:671)
> 	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.requestFailure(OperationSupport.java:651)
> 	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.assertResponseCode(OperationSupport.java:600)
> 	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.lambda$handleResponse$0(OperationSupport.java:560)
> 	at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:646)
> 	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
> 	at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2147)
> 	at io.fabric8.kubernetes.client.http.StandardHttpClient.lambda$completeOrCancel$10(StandardHttpClient.java:140)
> 	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863)
> 	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:841)
> 	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
> 	at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2147)
> 	at io.fabric8.kubernetes.client.http.ByteArrayBodyHandler.onBodyDone(ByteArrayBodyHandler.java:52)
> 	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863)
> 	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:841)
> 	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
> 	at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2147)
> 	at io.fabric8.kubernetes.client.vertx.VertxHttpRequest.lambda$null$1(VertxHttpRequest.java:122)
> 	at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:264)
> 	at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:246)
> 	at io.vertx.core.http.impl.HttpEventHandler.handleEnd(HttpEventHandler.java:76)
> 	at io.vertx.core.http.impl.HttpClientResponseImpl.handleEnd(HttpClientResponseImpl.java:250)
> 	at io.vertx.core.http.impl.Http1xClientConnection$StreamImpl.lambda$new$0(Http1xClientConnection.java:444)
> 	at io.vertx.core.streams.impl.InboundBuffer.handleEvent(InboundBuffer.java:255)
> 	at io.vertx.core.streams.impl.InboundBuffer.write(InboundBuffer.java:134)
> 	at io.vertx.core.http.impl.Http1xClientConnection$StreamImpl.handleEnd(Http1xClientConnection.java:708)
> 	at io.vertx.core.impl.EventLoopContext.lambda$execute$2(EventLoopContext.java:78)
> 	at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:174)
> 	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:167)
> 	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:470)
> 	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:569)
> 	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
> 	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
> 	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
> 	... 1 more
>
> 2023-08-23 18:07:05,182 ERROR [io.jav.ope.pro.eve.EventProcessor] (ReconcilerExecutor-keycloakcontroller-72) Exhausted retries for ExecutionScope{ resource id: ResourceID{name='cs-keycloak', namespace='keycloak-integration'}, version: 159092995}
>
> But it does not re-queue the KeyCloak CR reconciliation when this error
> happens.
> When the secret shows up in the namespace, KeyCloak operator stays silent
> and does not proceed. So I have to manually restart the operator pod to
> trigger a reconciliation
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/keycloak/keycloak/issues/22170#issuecomment-1692248639>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAGSVPDYYJ23FMYC634ADZ3XW6P4LANCNFSM6AAAAAA3AWOQ2A>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

--- a/operator/src/main/java/org/keycloak/operator/Constants.java
+++ b/operator/src/main/java/org/keycloak/operator/Constants.java
@@ -38,6 +38,7 @@ public final class Constants {
     public static final String KEYCLOAK_COMPONENT_LABEL = "operator.keycloak.org/component";
     public static final String KEYCLOAK_WATCHED_SECRET_HASH_ANNOTATION = "operator.keycloak.org/watched-secret-hash";
     public static final String KEYCLOAK_WATCHING_ANNOTATION = "operator.keycloak.org/watching-secrets";
+    public static final String KEYCLOAK_MISSING_SECRETS_ANNOTATION = "operator.keycloak.org/missing-secrets";
 
     public static final String DEFAULT_LABELS_AS_STRING = "app=keycloak,app.kubernetes.io/managed-by=keycloak-operator";
 

--- a/operator/src/main/java/org/keycloak/operator/controllers/KeycloakController.java
+++ b/operator/src/main/java/org/keycloak/operator/controllers/KeycloakController.java
@@ -130,7 +130,9 @@ public class KeycloakController implements Reconciler<Keycloak>, EventSourceInit
             updateControl = UpdateControl.updateStatus(kc);
         }
 
-        if (!status.isReady()) {
+        if (!status.isReady() || context.getSecondaryResource(StatefulSet.class)
+                .map(s -> s.getMetadata().getAnnotations().get(Constants.KEYCLOAK_MISSING_SECRETS_ANNOTATION))
+                .filter(Boolean::valueOf).isPresent()) {
             updateControl.rescheduleAfter(10, TimeUnit.SECONDS);
         }
 

--- a/operator/src/main/java/org/keycloak/operator/controllers/WatchedSecretsController.java
+++ b/operator/src/main/java/org/keycloak/operator/controllers/WatchedSecretsController.java
@@ -104,6 +104,8 @@ public class WatchedSecretsController implements Reconciler<Secret>, EventSource
     @Override
     public void annotateDeployment(List<String> desiredWatchedSecretsNames, Keycloak keycloakCR, StatefulSet deployment) {
         List<Secret> currentSecrets = fetchSecrets(desiredWatchedSecretsNames, keycloakCR.getMetadata().getNamespace());
+        deployment.getMetadata().getAnnotations().put(Constants.KEYCLOAK_MISSING_SECRETS_ANNOTATION,
+                Boolean.valueOf(currentSecrets.size() < desiredWatchedSecretsNames.size()).toString());
         deployment.getMetadata().getAnnotations().put(Constants.KEYCLOAK_WATCHING_ANNOTATION, desiredWatchedSecretsNames.stream().collect(Collectors.joining(";")));
         deployment.getSpec().getTemplate().getMetadata().getAnnotations().put(Constants.KEYCLOAK_WATCHED_SECRET_HASH_ANNOTATION, getSecretHash(currentSecrets));
     }

--- a/operator/src/main/java/org/keycloak/operator/controllers/WatchedSecretsController.java
+++ b/operator/src/main/java/org/keycloak/operator/controllers/WatchedSecretsController.java
@@ -44,6 +44,7 @@ import java.security.NoSuchAlgorithmException;
 import java.util.Arrays;
 import java.util.List;
 import java.util.Map;
+import java.util.Objects;
 import java.util.Optional;
 import java.util.stream.Collectors;
 
@@ -110,7 +111,8 @@ public class WatchedSecretsController implements Reconciler<Secret>, EventSource
     private List<Secret> fetchSecrets(List<String> secretsNames, String namespace) {
         return secretsNames.stream()
                 .map(n -> Optional.ofNullable(secrets).flatMap(cache -> cache.get(new ResourceID(n, namespace)))
-                        .orElseGet(() -> client.secrets().inNamespace(namespace).withName(n).require()))
+                        .orElseGet(() -> client.secrets().inNamespace(namespace).withName(n).get()))
+                .filter(Objects::nonNull)
                 .collect(Collectors.toList());
     }
 

--- a/operator/src/test/java/org/keycloak/operator/testsuite/integration/KeycloakDeploymentTest.java
+++ b/operator/src/test/java/org/keycloak/operator/testsuite/integration/KeycloakDeploymentTest.java
@@ -28,6 +28,7 @@ import io.fabric8.kubernetes.api.model.ServicePort;
 import io.fabric8.kubernetes.api.model.apps.StatefulSet;
 import io.fabric8.kubernetes.api.model.apps.StatefulSetBuilder;
 import io.fabric8.kubernetes.api.model.apps.StatefulSetSpecBuilder;
+import io.fabric8.kubernetes.client.dsl.Resource;
 import io.quarkus.logging.Log;
 import io.quarkus.test.junit.QuarkusTest;
 
@@ -92,6 +93,26 @@ public class KeycloakDeploymentTest extends BaseOperatorTest {
         k8sclient.resource(kc).delete();
         Awaitility.await()
                 .untilAsserted(() -> assertThat(k8sclient.apps().statefulSets().inNamespace(namespace).withName(deploymentName).get()).isNull());
+    }
+
+    @Test
+    public void testKeycloakDeploymentBeforeSecret() {
+        // CR
+        var kc = getTestKeycloakDeployment(true);
+        var deploymentName = kc.getMetadata().getName();
+        deployKeycloak(k8sclient, kc, false, false);
+
+        // Check Operator has deployed Keycloak and the statefulset exists, this allows for the watched secret to be picked up
+        Log.info("Checking Operator has deployed Keycloak deployment");
+        Resource<StatefulSet> stsResource = k8sclient.resources(StatefulSet.class).withName(deploymentName);
+        Resource<Keycloak> keycloakResource = k8sclient.resources(Keycloak.class).withName(deploymentName);
+        // expect no errors and not ready, which means we'll keep reconciling
+        Awaitility.await().untilAsserted(() -> {
+            assertThat(stsResource.get()).isNotNull();
+            Keycloak keycloak = keycloakResource.get();
+            CRAssert.assertKeycloakStatusCondition(keycloak, KeycloakStatusCondition.HAS_ERRORS, false);
+            CRAssert.assertKeycloakStatusCondition(keycloak, KeycloakStatusCondition.READY, false);
+        });
     }
 
     @Test

--- a/operator/src/test/java/org/keycloak/operator/testsuite/integration/WatchedSecretsTest.java
+++ b/operator/src/test/java/org/keycloak/operator/testsuite/integration/WatchedSecretsTest.java
@@ -87,8 +87,6 @@ public class WatchedSecretsTest extends BaseOperatorTest {
         var kc = getTestKeycloakDeployment(false);
         deployKeycloak(k8sclient, kc, true);
 
-        var prevRevision = getStatefulSet(kc).getStatus().getUpdateRevision();
-
         var dbSecret = getDbSecret();
 
         dbSecret.getData().put("username",

